File: TwoPointer/1004.max-consecutive-ones-iii.go
package leetcode

// @leet start
func longestOnes(nums []int, k int) int {
	var l, ret, cnt0 int
	for r, v := range nums {
		cnt0 += 1 - v
		// if 0 exceed k count
		for cnt0 > k {
			cnt0 -= 1 - nums[l]
			l++
		}

		ret = max(ret, r-l+1)
	}
	return ret
}

// @leet end

File: TwoPointer/1004.max-consecutive-ones-iii.py
from typing import List


# @leet start
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        l = res = count = 0
        for r, v in enumerate(nums):
            if v == 0:
                count += 1
            while count > k:
                if nums[l] == 0:
                    count -= 1
                l += 1
            cur_length = r - l + 1
            res = max(res, cur_length)

        return res


# @leet end

